test/validation/api/pktio/lso.c

```diff
@@ -0,0 +1,669 @@
+
+	iface_name[0] = getenv("ODP_PKTIO_IF0");
+	iface_name[1] = getenv("ODP_PKTIO_IF1");
+	num_ifaces = 1;
+
+	if (!iface_name[0]) {
+		printf("No interfaces specified, using default \"loop\".\n");
+		iface_name[0] = "loop";
+	} else if (!iface_name[1]) {
+		printf("Using loopback interface: %s\n", iface_name[0]);
+	} else {
+		num_ifaces = 2;
+		printf("Using paired interfaces: %s %s\n",
+		       iface_name[0], iface_name[1]);
+	}
+
+	if (pkt_pool_create() != 0) {
+		ODPH_ERR("Failed to create pool\n");
+		return -1;
+	}
+
+	/* Create pktios and associate input/output queues */
+	for (i = 0; i < num_ifaces; ++i) {
+		odp_pktio_t pktio;
+		const char *name = iface_name[i];
+
+		pktio = create_pktio(i, name, lso_pool);
+
+		if (pktio == ODP_PKTIO_INVALID) {
+			ODPH_ERR("Failed to open interface: %s\n", name);
+			return -1;
+		}
+
+		if (odp_pktout_queue(pktio, &pktios[i].pktout, 1) != 1) {
+			ODPH_ERR("Failed to get pktout queue: %s\n", name);
+			return -1;
+		}
+
+		if (odp_pktin_queue(pktio, &pktios[i].pktin, 1) != 1) {
+			ODPH_ERR("Failed to get pktin queue: %s\n", name);
+			return -1;
+		}
+	}
+
+	pktio_a = &pktios[0];
+	pktio_b = &pktios[1];
+	if (num_ifaces == 1)
+		pktio_b = pktio_a;
+
+	return 0;
+}
+
+int lso_suite_term(void)
+{
+	int i;
+	int ret = 0;
+
+	for (i = 0; i < num_ifaces; ++i) {
+		if (odp_pktio_close(pktios[i].hdl)) {
+			ODPH_ERR("Failed to close pktio: %s\n", pktios[i].name);
+			ret = -1;
+		}
+	}
+
+	if (odp_pool_destroy(lso_pool) != 0) {
+		ODPH_ERR("Failed to destroy pool\n");
+		ret = -1;
+	}
+
+	if (odp_cunit_print_inactive())
+		ret = -1;
+
+	return ret;
+}
+
+static int check_lso_custom(void)
+{
+	if (pktio_a->capa.lso.max_profiles == 0 || pktio_a->capa.lso.max_profiles_per_pktio == 0)
+		return ODP_TEST_INACTIVE;
+
+	if (pktio_a->capa.lso.proto.custom == 0 || pktio_a->capa.lso.mod_op.add_segment_num == 0)
+		return ODP_TEST_INACTIVE;
+
+	return ODP_TEST_ACTIVE;
+}
+
+static int check_lso_custom_restart(void)
+{
+	if (check_lso_custom() == ODP_TEST_INACTIVE)
+		return ODP_TEST_INACTIVE;
+
+	if (disable_restart && num_starts > 0)
+		return ODP_TEST_INACTIVE;
+
+	/* Run only one packet IO test case when interface restart is disabled */
+	num_starts++;
+
+	return ODP_TEST_ACTIVE;
+}
+
+static void lso_capability(void)
+{
+	/* LSO not supported when max_profiles is zero */
+	if (pktio_a->capa.lso.max_profiles == 0 || pktio_a->capa.lso.max_profiles_per_pktio == 0)
+		return;
+
+	CU_ASSERT(pktio_a->capa.lso.max_profiles >= pktio_a->capa.lso.max_profiles_per_pktio);
+	CU_ASSERT(pktio_a->capa.lso.max_packet_segments > 0);
+	/* At least 32 bytes of payload */
+	CU_ASSERT(pktio_a->capa.lso.max_payload_len >= 32);
+	/* LSO can create at least two segments */
+	CU_ASSERT(pktio_a->capa.lso.max_segments > 1);
+	/* LSO can copy at least Ethernet header to segments */
+	CU_ASSERT(pktio_a->capa.lso.max_payload_offset >= 14);
+
+	if (pktio_a->capa.lso.proto.custom)
+		CU_ASSERT(pktio_a->capa.lso.max_num_custom > 0);
```

Comment:
In this case `add_segment_num` or `add_payload_len` or `add_payload_offset` should also be supported.

Reply:
fixed